### ISSUE TITLE
Versioned API support

### ISSUE THREAD

**Issue by kissmikijr** [backstage/backstage#11027]
I'd like this to be more of a conversation starter about this topic, is it a needed feature? Does anybody else have this issue? The proposed solutions are just really quick ideas around the topic.

## Feature Suggestion
Usually, APIs are evolving during their lifetimes with different versions. In the current form, every version of an API spec will (and should) have its own API entity in backstage. However, all of these versions of APIs are actually tied to a component. From the POV of a service, it provides 1 API with different versions, but it is rendered right now as the service provides all the versions (which is technically true) but the user experience in finding the correct version can be bad if you have a lot of versions for your API.

I suggest that API specs could have a similar model like groups and users. Where I could see 1 "parent" API and I would be able to choose which version of that API I'd like to see.
example:
![image](https://user-images.githubusercontent.com/24729496/164707340-6b765ec1-a74d-4449-b5fd-3f73fc7b54cd.png)
with entities:
![image](https://user-images.githubusercontent.com/24729496/164711374-27bc208f-f78f-4f56-9365-ff6e7b03f73f.png)

## Possible Implementation
I think there could be a couple of ways to do this.
One idea is to introduce a new entity kind, stg like GroupApi. These could have a `spec.apis` array, where you should list your regular API entities. With this approach, it might be more clear to create frontend components for the GroupApis.

Another idea is to extend the current API entity with a `spec.versions` array, where you could list the other versions of this API spec. In this case, handling on the frontend might not be as clean, there would be not a clear distinction between regular API specs and GroupApi specs.

As a POC I was able to kinda hack around and create something like I mentioned above. For this, I used the API entity's existing `spec.owner` field. Where 1 API is actually owning other APIs and the top-level parent API is owned by a team/user. If you set up your APIs like this, then there is a possibility to create a frontend component that takes advantage of the `useRelatedEntities` hook to list all the entities that the current API owns. 
It is a hack since the official documentation states that the `spec.owner` field should be only user or groups. In this case, I used an `api` as an owner. 
In the current form API entities' `spec.definition` is a required field, however, the parent API will not have a direct definition so you'd need to use some dummy texts. 
An unsolved problem with this is handling the frontend for it, we could update the `ApiDefinitionCard` to handle the versioning, but it should handle the cases where the API is just a regular API and where it is a parent API.

**Comment by freben** [comment]
The catalog, in general, as you and @dkichler note above, _explicitly_ does not deal with versions. This includes APIs. The catalog is not meant to contain a perfect index of all versions of all artifacts of all variations of all software components in all slightly different environments - its entities are rather meant to be the higher level "ideas" of things; essentially at a granularity that humans talk about them.

For a service, the catalog does indeed not deal with the different deployed versions of that service. There may come a point where a new backwards-incompatible major version of the service is released, or when it's split into several smaller services, where you might consider also changing the entity structure to reflect that. But the main purpose of these entities is effectively to serve as "anchors" to tie functionality and relations onto, and to hold enough foreign key material to make sensible lookups in _other_ systems to find more granular data.

This philosophy translates to APIs as well. No matter how many patch/minor versions you make of an API, it's still effectively the same API in the eyes of the catalog. There may perhaps be a `myApiV1` and a `myApiV2` when you are at the point that they are majorly diverging so much that they effectively are completely different APIs perhaps even served by different services.

For this reason, I think that the `definition` field may have been a mistake in the first place. It has led to this type of thinking that it has to perfectly reflect an exact version of the API. But actually, it was only ever meant to be lightly informative, probably of a recent version of the API, to make it easy for engineers to quickly have a peek at it and interact with it. But I think that that use case might have been better served in other ways than to store the actual schema in the entity itself. So I vote to actually remove - or make optional - this field in a future version of the API kind. Let's see if that pans out - it's just my personal opinion :)

**Comment by freben** [comment]
So to clarify, I want to push back against the formulation

> every version of an API spec will (and should) have its own API entity in backstage

and encourage people to NOT do that :)

**Comment by dkichler-improving** [comment]
I agree @freben - I think versioning should be a part of the entity model itself, whereby an entity can have a traceable lineage of versions all associated with the top-level entity.

**Comment by dkichler-improving** [comment]
Apologies, replied before I read your initial reply explaining the reasoning on why versioning is _not_ part of the entity model.  It makes sense - bringing versioning into the model would come with a massive amount of complexity.  Certainly there is still plenty of value in a "lightly informative, probably of a recent version of the API" (or service).

**Comment by cmars** [comment]
> So to clarify, I want to push back against the formulation
> 
> > every version of an API spec will (and should) have its own API entity in backstage
> 
> and encourage people to NOT do that :)

At Snyk we're using a rolling [date- and stability-based API versioning scheme](https://github.com/snyk/sweater-comb/blob/main/docs/principles/version.md#api-versioning). It's not so much `myAPIv1` and `myAPIv2` but `myAPI?version=2022-06-08~experimental` and `myAPI?version=2022-05-05~ga`. API consumers can choose the version they want to interact with.

Not representing this in Backstage would be very unfortunate, as it's already such a useful catalog of services and their APIs. Especially APIs; our instance is rendering OpenAPI documentation using Swagger UI.

> its entities are rather meant to be the higher level "ideas" of things; essentially at a granularity that humans talk about them.

My counter-argument to this (I'm admittedly very biased here!): humans also need to talk about changes in components over time. In a distributed system that's seeing significant use, it's more likely that an old version of a component deprecates and sunsets and a new version emerges and takes over responsibility as a gradual process. During this transition time, I would expect that consumers of that component will need information about both the old and the new version.

It's entirely possible to represent a linear timeline of versions with distinct entity names -- it's just that the UI wouldn't represent this very well, what seems to be missing then is a way of "grouping" these multiple versions. If there were a metadata field or convention that could add this dimension, it would be a quite powerful extension to the existing data model.

I'm not entirely clear on the complexities this would bring though -- what challenges would you face in adding this dimension?

**Comment by freben** [comment]
Thanks for your detailed reply!

I do appreciate that there are underlying needs to find more detailed information along a lot of axes in relation to entities. and indeed we do expose a very rich set of information in our views, and link out to extended details in different subsystems or cloud provider pages etc. So the _end user_ experience should definitely be to have all of the information they need at their fingertips. Or in some cases, a cleverly digested summary of it, linking out to the full actual data elsewhere.

So we want to achieve the same thing; the main question is, should the _implementation_ of that need be to specifically put that data into the catalog, or to have it elsewhere. After all, the catalog is just a low level backend service, not a user facing system.

My view on the catalog is heavily centered around the engineer end user experience. As in, the people that go through the chore of actually sourcing data into the catalog. In all honesty, asking them to maintain yaml files by hand and getting all of the data and relations right, is a pretty tall order, one that I'm acutely aware of. :) There was a pretty serious balancing act going on in the design there, gauging power and expressiveness vs customisability vs maintenance cost etc. We want to burden engineers as little as possible, letting them express the most vital pieces of information only where necessary, and then get out of their way.

I think it's safe to say that it'd be desirable to avoid having end users (again, not frontend users, but yaml users, so to speak) to have to sit and copy-paste largish yaml blocks and carefully hand tweak them and invent names for them, for every API change, basically just to appease Backstage. And then when their service declared that it depends on ApiV1.1.2, but later they tweak their code to talk to ApiV1.1.3 - now they have to remember to update their dependency in the yaml file to the other version of that entity, and oh the other team forgot to make an entity of that version so we need to contact them ... etc.

It'd probably turn out to be frustrating. We want the experience to be delightful and ultimately increasing productivity. A feeling of a small investment that gets them big value by unlocking the Backstage ecosystem, as it were. As you know, "selling" a system like this onto an org full of busy engineers, is largely a social engineering exercise as much as a technical one.

So the alternative to that would be some form of automation. Mirroring data from an external system into the catalog - for example, for library type components, you could go to artifactory and ensure that there's an entity present for every single actual published version of the library. But then the question becomes - why? Why mirror that data in the first place? How do you fill in things like the owner field on those spidered things? There ARE some cases where you do want to mirror things, as we do for example for users and groups, but that should be limited to a very, VERY small set of things where you know that the implications of doing so are actually worth the (tangible) cost and where doing so brings large enough value to the Backstage ecosystem or other consumers of the catalog - where they conceivably could instead have looked up that same info themselves from the authoritative system if they needed it.

Remember, putting something into the catalog implies a number of things. One of them is that they bloat search and similar discovery use cases. As a user, you're in an incident scenario, and you have a hunch that the service `foo` that you think you transitively depend on may be the culprit. So you go to search and start typing `foo` and get eighteen hits. Hm, now what? :) You yourself alluded to grouping, which would get around this, but I fear that such a feature would take some significant effort and affects more systems such as the frontend, scaffolding, search, etc etc etc. Another cost is just the plain suitability aspect. In fairness I think that the catalog is pretty custom built around its current scope and use case, and is a pretty poor performer when it comes to - as an example - being an artifact index, or an API gateway/index, etc. There are perfectly suitable products for those other needs, and we want to leverage those and not reinvent what they do, much in the same way that we generally prefer frontend plugins that show the critical most useful pieces of info but then link out to the third party system instead of reinventing their entire UI.

So winding back a bit. What we focus on, is to have these "hook" entities that represent the high level concepts, and then we very carefully craft the web of connections to other authorities that have the rest of the info.

Now, on that topic. It may be that you desire things to be in the catalog because it implies ease of consumption (from the REST API standpoint basically). And that is understandable and a common thing that we hear. Even at Spotify, my feeling has for the longest time been that we are great at building islands of functionality but could be a bit better at focusing on the edges _between_ those pieces. And that's where graphql enters the picture. Internally we have a legacy graphql layer available to the frontend, that mirrors the catalog and many many other systems. From the entity nodes you can resolve a ton of different things, not by asking the catalog, but by descending into those external systems looking data up by the entity ref. This is where we want to go with the open source product as well, but as you may be aware we have just an embryo of it and haven't yet had time to focus on making it into a proper product. This is a direction we believe in; an "edge" linking layer on top of the catalog, rather than putting any and all data INTO the catalog itself.

Does that make sense? I got a bit rambling there, please feel free to ask for clarification or offer disagreement :) I like the discourse.

**Comment by webark** [comment]
@Rugvip I would like to explore that option of api definitions not sitting directly in the api entity as well. 

Some people from our organization who have started to adopt backstage assumed that API definitions were more similarly dealt with as TechDocs rather than a component or resource. I wonder if there could be some benefit of having a base concept that tech docs and api definitions shared, which could deal with the shared need of things like addons, versioning, generation, publishing etc.

**Comment by Rugvip** [comment]
@webark hmm, possibly. There's definitely a bit of overlap, but also some pretty large differences. Could also be that it's more of a common service situation where they're available for reuse by all plugins, for example like UrlReaders. Breaking things into smaller single-purpose atoms that then get composed back together tends to be a bit easier to evolve.

**Comment by moley** [comment]
Hi, I think versioned API support would be a really sensible feature. Do you have any update regarding this feature?

**Comment by swestner-loyalty** [comment]
I would be happy with environmental support. The use case here being that we have a deployment process that goes through a number of environments (e.g. test, uat, prod) and that the documentation in one of these environments is always out of sync. 

For example, we update an API and spec, as soon as it enters the pipeline, backstage it breaks prod documentation. If we do not update specs until we push to prod, then the lower environments end up broken

Having something like a lifecycle tag preserve docs in the experimental stage separately from production would allow the pushing of updates through lower environments without changing production docs until ready. Then automating updates to the  catalog files to put in production lifecycles when pushing to prod 
 would be trivial
 
 At least that would be intuitively how I would expect lifecycles to work, and it would get at least part way to the goal of the OP without the kludge of becoming a SCM tool

**Comment by Rugvip** [comment]
Bit of extra context over in #20209

**Comment by tkrop** [comment]
At Zalando we are also struggling with the same shortcoming of the API kind and API docs plugin.

Our legacy API portal treats APIs as something that has a history of semantic versions - and if users forget to update the version add new revisions. We automatically track every API that is deployed - in tests and production systems -, and keep the API definition in our database forever as part of an API history.

![image](https://github.com/backstage/backstage/assets/18184975/d7437905-bafd-4ab3-a720-107e59aa1ee7)

Since the API model in Backstage binds to a single API spec, we are currently forced to create one API entity per API revisions instead of one API entity per API history, i.e. a collection of related API revisions. This creates two problems:

1. We need to create a unique API name for every revision, which leads to unstable names and links from API perspective, that users cannot reference in APIs in discussions, and worse
2. We need to remove inactive API revision to prevent overloading of the catalog as well as the human perception, because the number of API revisions has grown to multiple hundred of thousands in our company.

To represent our legacy API portal information in Backstage compatible with the current API docs plugin, we suggest to add an optional `#/spec/versions`-array that enables to bind the API definition to a distinct version while at the same time relax `#/spec/definition` to be optional as in the following example:

```yaml
apiVersion: backstage.io/v1alpha1
kind: API
metadata:
  name: api-repository-api
  description: API Repository API (external)
specs:
  type: openapi
  lifecycle: production
  owner: api-management-team
  system: api-repository
  versions:
    - version: 0.2.7
      revision: 1 # optional suggestion
      href: https://backstage.zalando.net/api-revisions/api-repository-api:0.2.7-1
    - version: 0.2.6
      revision: 2 # optional suggestion
      href: https://backstage.zalando.net/api-revisions/api-repository-api:0.2.6-2
    - ...
```

The data structure may contain also a `definition` on each version, however, resolving the API as `yaml` or `json` from a hypertext reference seems to be more preferable for us since it provides a cleaner separation of concerns. With this information the API definition view of the plugin can now easily show a `version`-menu to select another version than the last (active). To complete the feature, the API definition view can also support a version suffix `:{version}-{revision}` to directly point to a distinct API version.

Another extension would be to provide a `state` for each version that allows to distinguish between API revisions that are provided by staging and/or production services or have been retired.

**Comment by benjdlambert** [comment]
Wonder if we should look at decoupling the inline schema from the `API` entity, and having the definition and versions resolved in a different way. Something similar to the above, but maybe it's taken care of by an entirely different service rather than this stored in the Catalog? Not sure.

**Comment by kiranpatel11** [comment]
other alternative is to use the purpose built tools like [apicurio](https://www.apicur.io/registry/) and integrate it via the plugin.

**Comment by tridentsx** [comment]
Backstage assumes that you running a SaaS service! What if your developping sw as a vendor, that is used by 100's of customers, not all of them will have same version running at the same time in production. Still support and design maintenance would be delivered, meaning that S/W developpers daily needs to lookup things in different versions of all entities. I understand for Spotify this is not an issue but any S/W vendor with large customer base it is.

**Comment by dominikjeske** [comment]
Are there any plans to implement this feature?

**Comment by Edje-C** [comment]
We don't have intentions to pick this up, it is not currently on our roadmap. 
This is what we think needs to be done
https://github.com/backstage/backstage/issues/11027#issuecomment-1752843513

If someone would like to contribute, PRs are welcome 🙏

**Comment by chicobento** [comment]
> We don't have intentions to pick this up, it is not currently on our roadmap. This is what we think needs to be done [#11027 (comment)](https://github.com/backstage/backstage/issues/11027#issuecomment-1752843513)
> 
> If someone would like to contribute, PRs are welcome 🙏

That would solve for APIs (which is what this issue title is about), but that wouldn't solve the broader catalog versioning gap as mentioned [here](https://github.com/backstage/backstage/issues/11027#issuecomment-1145366942). It would be nice to have it solved more generically for other entities as well, such as docs, components and entities relations, etc...

**Comment by webark** [comment]
I think that general catalog versioning is a separate subject then api spec, docs, etc. Anything that shares the same "support document" like specs and docs should get piped through the same process i see, and be able to store multiple times. (a versioned api spec or doc is just multiple specs associated by a specified key after all)

**Comment by chicobento** [comment]
> With this information the API definition view of the plugin can now easily show a `version`-menu to select another version than the last (active). To complete the feature, the API definition view can also support a version suffix `:{version}-{revision}` to directly point to a distinct API version.

What about the provides/consumes relationships ? Ideally components should be able to specify which version of the API they produce or consume.

> Another extension would be to provide a state for each version that allows to distinguish between API revisions that are provided by staging and/or production services or have been retired.

This would collide with spec.lifecycle field.

I was thinking more on a "common to all kinds" version field that would specify the entity version and allow adding relationships to specific versions, or the generic/default version (latest). Different versions would also be able to define different labels, annotations, lifecycle, relationships and such. The version dropdown field would be conditionally added to every entity page that has multiple versions.
Off course, this would add more complexity on the descriptors which would also have to be taken into consideration.

**Comment by webark** [comment]
Versioning the entire catalog is a paradigm shift in how the system operates, and the data is stored, grouped, and aggregated. Closing the loop between your api spec, components, and tracing service Id say would be a better fit if that's the need of your organization. Then you can pull those kind of fields and relationships dynamically based on the tracing data.

**Comment by sennyeya** [comment]
I wonder if a combination of metadata on annotations* and decoupling the API spec from the API entity would solve this nicely. You could put version on your provides/consumes relationship with a concise 'namespace:kind/name@version' or something and then that could be resolved to pull the correct version. We'd probably want to be clear that this is content versioning not entity versioning. Rereading the above thread, I think freben's comment above rings true for versioning entities in the catalog.

Using metadata on relations and a common 'VersionResolverService' might also make this usable for tech docs and content versioned stuff?

* metadata on relations is something I'm working on but currently isn't possible.

**Comment by adminy** [comment]
Wow, not even a workaround, sounds terrific, time to ditch backstage.

**Comment by moley** [comment]
I think this feature would raise popularity of backstage even more, because I think many companies, which do not work against one master branch would leverage

**Comment by webark** [comment]
yea. We currently generate versions of our spec when we build services, and we store them on S3 and have a service we can query to get them.

So we were able to just overwrite the api spec render to provide a dropdown, and add a versions block to the api spec.

I still feel the long term solution is to treat docs and specs the same. The same in that they are "associated assets" that get generated via the same type of build triggers that are used to tech docs for instance, and can publish them in similar ways. Each time they build, a new one is generated anyway. versioning just becomes connecting the existing versions that are generated using the techdocs cli, or in the real time builder that is available for tech docs already.

(this could also be used for SBOMs for example too. but don't want to 🧵 too much)

**Comment by jjwedemyer** [comment]
We built a custom Kind which acts as an addition to the API kind. it has a kv map from version to file.
And displaying it has a dropdown to select version you wanna look at.

**Comment by chicobento** [comment]
> We built a custom Kind which acts as an addition to the API kind. it has a kv map from version to file.
> And displaying it has a dropdown to select version you wanna look at.

Cool. In our case we are following same strategy proposed by @tkrop in https://github.com/backstage/backstage/issues/11027#issuecomment-1748608197: still using API kind but extending the `spec` field so that we can display the dropdown

**Comment by adminy** [comment]
@webark @jjwedemyer any packages published that others could use? Thanks
